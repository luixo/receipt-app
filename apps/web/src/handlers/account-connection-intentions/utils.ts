import { TRPCError } from "@trpc/server";
import type { z } from "zod";

import { ACCOUNT_CONNECTIONS_INTENTIONS } from "~db/consts";
import type { AccountsId, UsersId } from "~db/models";
import type { Database } from "~db/types";
import type { emailSchema } from "~web/handlers/validation";

export const addConnectionIntention = async (
	database: Database,
	fromAccountId: AccountsId,
	user: { name: string },
	toEmail: z.infer<typeof emailSchema>,
	asUserId: UsersId,
) => {
	const targetAccount = await database
		.selectFrom("accounts")
		.select(["id", "avatarUrl"])
		.where("email", "=", toEmail.lowercase)
		.executeTakeFirst();
	if (!targetAccount) {
		throw new TRPCError({
			code: "NOT_FOUND",
			message: `Account with email "${toEmail.original}" does not exist.`,
		});
	}
	const connectedUser = await database
		.selectFrom("accounts")
		.where("accounts.email", "=", toEmail.lowercase)
		.innerJoin("users", (qb) =>
			qb
				.onRef("users.connectedAccountId", "=", "accounts.id")
				.on("users.ownerAccountId", "=", fromAccountId),
		)
		.select("users.name")
		.executeTakeFirst();
	if (connectedUser) {
		throw new TRPCError({
			code: "CONFLICT",
			message: `Account with email "${toEmail.original}" is already connected to user "${connectedUser.name}".`,
		});
	}
	const viceVersaIntention = await database
		.selectFrom("accountConnectionsIntentions")
		.select("userId")
		.where((eb) =>
			eb.and({
				accountId: targetAccount.id,
				targetAccountId: fromAccountId,
			}),
		)
		.executeTakeFirst();
	if (viceVersaIntention) {
		await database.transaction().execute(async (tx) => {
			await tx
				.updateTable("users")
				.set({ connectedAccountId: fromAccountId })
				.where((eb) =>
					eb.and({
						ownerAccountId: targetAccount.id,
						id: viceVersaIntention.userId,
					}),
				)
				.executeTakeFirst();
			await tx
				.updateTable("users")
				.set({ connectedAccountId: targetAccount.id })
				.where((eb) =>
					eb.and({
						ownerAccountId: fromAccountId,
						id: asUserId,
					}),
				)
				.executeTakeFirst();
			await tx
				.deleteFrom("accountConnectionsIntentions")
				.where((eb) =>
					eb.and({
						accountId: targetAccount.id,
						targetAccountId: fromAccountId,
					}),
				)
				.executeTakeFirst();
		});
		return {
			account: {
				id: targetAccount.id,
				email: toEmail.lowercase,
				avatarUrl: targetAccount.avatarUrl || undefined,
			},
			connected: true,
			user: {
				name: user.name,
			},
		};
	}
	try {
		await database
			.insertInto("accountConnectionsIntentions")
			.values({
				accountId: fromAccountId,
				targetAccountId: targetAccount.id,
				userId: asUserId,
				createdAt: new Date(),
			})
			.executeTakeFirst();
		return {
			account: {
				id: targetAccount.id,
				email: toEmail.lowercase,
				avatarUrl: targetAccount.avatarUrl || undefined,
			},
			connected: false,
			user: {
				name: user.name,
			},
		};
	} catch (e) {
		// Could be like `duplicate key value violates unique constraint "..."`
		const message = String(e);
		if (
			message.includes(ACCOUNT_CONNECTIONS_INTENTIONS.CONSTRAINTS.ACCOUNT_PAIR)
		) {
			const existingIntention = await database
				.selectFrom("accountConnectionsIntentions")
				.where((eb) =>
					eb.and({
						accountId: fromAccountId,
						targetAccountId: targetAccount.id,
					}),
				)
				.select(["userId"])
				.executeTakeFirstOrThrow();
			const existingUser = await database
				.selectFrom("users")
				.select("name")
				.where("id", "=", existingIntention.userId)
				.executeTakeFirstOrThrow();
			throw new TRPCError({
				code: "CONFLICT",
				message: `You already has intention to connect to "${toEmail.original}" as user "${existingUser.name}".`,
			});
		}
		if (
			message.includes(ACCOUNT_CONNECTIONS_INTENTIONS.CONSTRAINTS.USER_PAIR)
		) {
			throw new TRPCError({
				code: "CONFLICT",
				message: `You already has intention to connect to user "${user.name}".`,
			});
			// This is probably a c8 bug
			/* c8 ignore next */
		}
		/* c8 ignore next 2 */
		throw e;
	}
};
