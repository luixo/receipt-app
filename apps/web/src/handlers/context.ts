import { type H3Event, createEvent } from "@tanstack/react-start/server";
import type { inferProcedureBuilderResolverOptions } from "@trpc/server";
import type { TRPCRequestInfo } from "@trpc/server/http";
import type { IncomingMessage, ServerResponse } from "node:http";

import type { Database } from "~db/database";
import type { TestContext } from "~tests/backend/utils/test";
import type { authProcedure } from "~web/handlers/trpc";
import type { CacheDbOptions } from "~web/providers/cache-db";
import type { EmailOptions } from "~web/providers/email";
import type { ExchangeRateOptions } from "~web/providers/exchange-rate";
import type { Logger } from "~web/providers/logger";
import type { S3Options } from "~web/providers/s3";

type TestContextPicks = Pick<TestContext, "getSalt" | "getUuid"> & {
	database: Database;
	logger: Logger;
	emailOptions: EmailOptions;
	cacheDbOptions: CacheDbOptions;
	exchangeRateOptions: ExchangeRateOptions;
	s3Options: S3Options;
};

export type NetContext = {
	event: H3Event;
	info: TRPCRequestInfo;
};

export type UnauthorizedContext = NetContext & TestContextPicks;

export type AuthorizedContext = inferProcedureBuilderResolverOptions<
	typeof authProcedure
>["ctx"];

export const createContext = (
	{
		req,
		res,
		info,
	}: { req: IncomingMessage; res: ServerResponse; info: TRPCRequestInfo },
	testContext: TestContextPicks,
): UnauthorizedContext => ({
	info,
	event: createEvent(req, res),
	...testContext,
});
