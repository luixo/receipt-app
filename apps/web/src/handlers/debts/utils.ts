import type { Database, DebtsId, SimpleInsertObject } from "~db";
import { nonNullishGuard } from "~utils";

export const upsertAutoAcceptedDebts = async (
	database: Database,
	debts: (SimpleInsertObject<"debts"> & { isNew: boolean })[],
) => {
	const fetchedDebts = await database
		.selectFrom("debts")
		.where((eb) =>
			eb.or(
				debts.map((debt) =>
					!debt.isNew
						? eb.and({
								id: debt.id,
								ownerAccountId: debt.ownerAccountId,
						  })
						: debt.receiptId
						? eb.and({
								ownerAccountId: debt.ownerAccountId,
								userId: debt.userId,
								receiptId: debt.receiptId,
						  })
						: eb("debts.ownerAccountId", "is", null),
				),
			),
		)

		.select([
			"debts.id",
			"debts.ownerAccountId",
			"debts.userId",
			"debts.receiptId",
		])
		.execute();
	const counterpartyDebts = debts.map((debt) =>
		fetchedDebts.find((fetchedDebt) => {
			if (debt.isNew) {
				if (!debt.receiptId || !fetchedDebt.receiptId) {
					return false;
				}
				return (
					fetchedDebt.ownerAccountId === debt.ownerAccountId &&
					fetchedDebt.userId === debt.userId &&
					fetchedDebt.receiptId === debt.receiptId
				);
			}
			return (
				fetchedDebt.ownerAccountId === debt.ownerAccountId &&
				fetchedDebt.id === debt.id
			);
		}),
	);
	const existentDebts = debts
		.map((nextDebt, index) =>
			counterpartyDebts[index]
				? ([nextDebt, counterpartyDebts[index]!] as const)
				: null,
		)
		.filter(nonNullishGuard);
	const nonExistentDebts = debts
		.map((nextDebt, index) => (counterpartyDebts[index] ? null : nextDebt))
		.filter(nonNullishGuard);
	const [newDebts, ...updatedDebts] = await Promise.all([
		nonExistentDebts.length === 0
			? ([] as { id: DebtsId }[])
			: database
					.insertInto("debts")
					.values(nonExistentDebts.map(({ isNew, ...debt }) => debt))
					.returning(["debts.id"])
					.execute(),
		...existentDebts.map(([nextDebt, currentDebt]) =>
			database
				.updateTable("debts")
				.set({
					amount: nextDebt.amount,
					timestamp: nextDebt.timestamp,
					currencyCode: nextDebt.currencyCode,
					receiptId: nextDebt.receiptId,
					lockedTimestamp: nextDebt.lockedTimestamp,
				})
				.where((eb) =>
					eb.and({
						id: currentDebt.id,
						ownerAccountId: currentDebt.ownerAccountId,
						userId: currentDebt.userId,
					}),
				)
				.returning(["debts.id", "debts.receiptId"])
				.executeTakeFirst(),
		),
	]);
	return { newDebts, updatedDebts };
};
