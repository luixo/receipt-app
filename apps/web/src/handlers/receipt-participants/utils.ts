import { TRPCError } from "@trpc/server";
import type { Selection } from "kysely";
import type { z } from "zod";

import type { Database, ReceiptsSelectExpression } from "~web/db";
import type { AccountsId, ReceiptsId, UsersId } from "~web/db/models";
import type { ReceiptsDatabase } from "~web/db/types";
import type { assignableRoleSchema } from "~web/handlers/validation";

export const getReceiptParticipant = <
	SE extends ReceiptsSelectExpression<"receiptParticipants">,
>(
	database: Database,
	userId: UsersId,
	receiptId: ReceiptsId,
	selectExpression: SE[],
): Promise<
	Selection<ReceiptsDatabase, "receiptParticipants", SE> | undefined
> =>
	database
		.selectFrom("receiptParticipants")
		.where((eb) => eb.and({ receiptId, userId }))
		.select(selectExpression)
		.executeTakeFirst();

export const addReceiptParticipants = async (
	database: Database,
	receiptId: ReceiptsId,
	receiptOwnerId: AccountsId,
	receiptOwnerEmail: string,
	usersToAdd: [UsersId, z.infer<typeof assignableRoleSchema>][],
) => {
	const userIds = usersToAdd.map(([id]) => id);
	const users = await database
		.selectFrom("users")
		.where("users.ownerAccountId", "=", receiptOwnerId)
		.select("users.id")
		.where("users.id", "in", userIds)
		.execute();
	if (users.length !== userIds.length) {
		const missedUserIds = userIds.filter(
			(userId) => !users.some((user) => user.id === userId),
		);
		throw new TRPCError({
			code: "NOT_FOUND",
			message: `${missedUserIds.length === 1 ? "User" : "Users"} ${missedUserIds
				.map((id) => `"${id}"`)
				.join(", ")} ${
				missedUserIds.length === 1 ? "does" : "do"
			} not exist or ${
				missedUserIds.length === 1 ? "is" : "are"
			} not owned by "${receiptOwnerEmail}".`,
		});
	}
	const receiptParticipants = await database
		.selectFrom("receiptParticipants")
		.where((eb) => eb("receiptId", "=", receiptId).and("userId", "in", userIds))
		.select(["userId"])
		.execute();
	if (receiptParticipants.length !== 0) {
		throw new TRPCError({
			code: "CONFLICT",
			message: `${
				receiptParticipants.length === 1 ? "User" : "Users"
			} ${receiptParticipants
				.map(({ userId }) => `"${userId}"`)
				.join(", ")} already ${
				receiptParticipants.length === 1 ? "participates" : "participate"
			} in receipt "${receiptId}".`,
		});
	}
	const result = await database
		.insertInto("receiptParticipants")
		.values(
			usersToAdd.map(([id, role]) => ({
				receiptId,
				userId: id,
				role: receiptOwnerId === id ? "owner" : role,
				added: new Date(),
			})),
		)
		.returning(["added", "userId"])
		.execute();
	return usersToAdd.map(([id, role]) => {
		const addedDatum = result.find(({ userId }) => userId === id)!;
		return {
			id,
			added: addedDatum.added,
			role: receiptOwnerId === id ? ("owner" as const) : role,
		};
	});
};
