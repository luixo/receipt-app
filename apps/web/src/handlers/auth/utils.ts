import { TRPCError } from "@trpc/server";

import { DAY } from "~utils";
import type { AccountsId, SessionsSessionId } from "~web/db/models";
import { generateConfirmEmailEmail } from "~web/email/utils";
import type { UnauthorizedContext } from "~web/handlers/context";
import { getEmailClient } from "~web/providers/email";

// How long a session should last
export const SESSION_EXPIRATION_DURATION = 30 * DAY;
// How long session should be intact before we update it
export const SESSION_SHOULD_UPDATE_EVERY = 2 * DAY;

export const getExpirationDate = () =>
	new Date(Date.now() + SESSION_EXPIRATION_DURATION);

export const createAuthorizationSession = async (
	ctx: UnauthorizedContext,
	accountId: AccountsId,
) => {
	const uuid: SessionsSessionId = ctx.getUuid();
	const expirationDate = getExpirationDate();
	await ctx.database
		.insertInto("sessions")
		.values({
			accountId,
			sessionId: uuid,
			expirationTimestamp: expirationDate,
		})
		.executeTakeFirst();
	return {
		authToken: uuid,
		expirationDate,
	};
};

export const sendVerificationEmail = async (
	ctx: UnauthorizedContext,
	email: string,
	token: string,
) => {
	if (!ctx.emailOptions.active) {
		throw new TRPCError({
			code: "FORBIDDEN",
			message: "Currently email resend is not supported",
		});
	}
	try {
		await getEmailClient(ctx).send({
			address: email,
			subject: "Confirm email in Receipt App",
			body: generateConfirmEmailEmail(ctx, token),
		});
	} catch (e) {
		throw new TRPCError({
			code: "INTERNAL_SERVER_ERROR",
			message: `Something went wrong: ${
				/* c8 ignore next */
				e instanceof Error ? e.message : String(e)
			}`,
		});
	}
};
