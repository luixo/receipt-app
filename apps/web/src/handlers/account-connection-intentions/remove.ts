import { TRPCError } from "@trpc/server";
import { z } from "zod";

import { authProcedure } from "~web/handlers/trpc";
import { accountIdSchema } from "~web/handlers/validation";

export const procedure = authProcedure
	.input(
		z.strictObject({
			targetAccountId: accountIdSchema,
		}),
	)
	.mutation(async ({ ctx, input }) => {
		const { database } = ctx;
		const intention = await database
			.selectFrom("accountConnectionsIntentions")
			.select(["accountId", "targetAccountId"])
			.where((eb) =>
				eb.and({
					accountId: ctx.auth.accountId,
					targetAccountId: input.targetAccountId,
				}),
			)
			.executeTakeFirst();
		if (!intention) {
			throw new TRPCError({
				code: "NOT_FOUND",
				message: `Intention for account id "${input.targetAccountId}" not found.`,
			});
		}
		await database
			.deleteFrom("accountConnectionsIntentions")
			.where((eb) =>
				eb.and({
					accountId: intention.accountId,
					targetAccountId: intention.targetAccountId,
				}),
			)
			.execute();
	});
