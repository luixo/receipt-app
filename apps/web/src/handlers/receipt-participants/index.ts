import { procedure as add } from "./add";
import { procedure as remove } from "./remove";
import { procedure as update } from "./update";

export const router = {
	add,
	remove,
	update,
};
