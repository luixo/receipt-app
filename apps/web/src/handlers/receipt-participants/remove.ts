import { TRPCError } from "@trpc/server";
import { z } from "zod/v4";

import { getReceiptParticipant } from "~web/handlers/receipt-participants/utils";
import { authProcedure } from "~web/handlers/trpc";
import { receiptIdSchema, userIdSchema } from "~web/handlers/validation";

export const procedure = authProcedure
	.input(
		z.strictObject({
			receiptId: receiptIdSchema,
			userId: userIdSchema,
		}),
	)
	.mutation(async ({ input, ctx }) => {
		const { database } = ctx;
		const receipt = await database
			.selectFrom("receipts")
			.select(["ownerAccountId"])
			.where("id", "=", input.receiptId)
			.limit(1)
			.executeTakeFirst();
		if (!receipt) {
			throw new TRPCError({
				code: "NOT_FOUND",
				message: `Receipt "${input.receiptId}" does not exist.`,
			});
		}
		if (receipt.ownerAccountId !== ctx.auth.accountId) {
			throw new TRPCError({
				code: "FORBIDDEN",
				message: `Not enough rights to remove participant from receipt "${input.receiptId}".`,
			});
		}
		const user = await database
			.selectFrom("users")
			.select("ownerAccountId")
			.where("id", "=", input.userId)
			.limit(1)
			.executeTakeFirst();
		if (!user) {
			throw new TRPCError({
				code: "NOT_FOUND",
				message: `User "${input.userId}" does not exist.`,
			});
		}
		if (user.ownerAccountId !== ctx.auth.accountId) {
			throw new TRPCError({
				code: "FORBIDDEN",
				message: `User "${input.userId}" is not owned by "${ctx.auth.email}".`,
			});
		}
		const receiptParticipant = await getReceiptParticipant(
			database,
			input.userId,
			input.receiptId,
			["userId"],
		);
		if (!receiptParticipant) {
			throw new TRPCError({
				code: "CONFLICT",
				message: `User "${input.userId}" does not participate in receipt "${input.receiptId}".`,
			});
		}
		await database.transaction().execute(async (tx) => {
			await tx
				.deleteFrom("receiptItemConsumers")
				.where((eb) =>
					eb("userId", "=", input.userId).and("itemId", "in", (ebb) =>
						ebb
							.selectFrom("receiptItems")
							.where("receiptId", "=", input.receiptId)
							.select("id"),
					),
				)

				.executeTakeFirst();
			await tx
				.deleteFrom("receiptParticipants")
				.where((eb) =>
					eb.and({
						receiptId: input.receiptId,
						userId: input.userId,
					}),
				)
				.executeTakeFirst();
		});
	});
