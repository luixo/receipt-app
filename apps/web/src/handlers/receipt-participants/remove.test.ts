import { faker } from "@faker-js/faker";
import { describe } from "vitest";

import { createAuthContext } from "~tests/backend/utils/context";
import {
	insertAccount,
	insertAccountWithSession,
	insertItemParticipant,
	insertReceipt,
	insertReceiptItem,
	insertReceiptParticipant,
	insertUser,
} from "~tests/backend/utils/data";
import {
	expectDatabaseDiffSnapshot,
	expectTRPCError,
	expectUnauthorizedError,
} from "~tests/backend/utils/expect";
import { test } from "~tests/backend/utils/test";
import { t } from "~web/handlers/trpc";

import { procedure } from "./remove";

const createCaller = t.createCallerFactory(t.router({ procedure }));

describe("receiptParticipants.remove", () => {
	describe("input verification", () => {
		expectUnauthorizedError((context) =>
			createCaller(context).procedure({
				receiptId: faker.string.uuid(),
				userId: faker.string.uuid(),
			}),
		);

		describe("receiptId", () => {
			test("invalid", async ({ ctx }) => {
				const { sessionId } = await insertAccountWithSession(ctx);
				const caller = createCaller(createAuthContext(ctx, sessionId));
				await expectTRPCError(
					() =>
						caller.procedure({
							receiptId: "not-a-uuid",
							userId: faker.string.uuid(),
						}),
					"BAD_REQUEST",
					`Zod error\n\nAt "receiptId": Invalid uuid`,
				);
			});
		});

		describe("userId", () => {
			test("invalid", async ({ ctx }) => {
				const { sessionId } = await insertAccountWithSession(ctx);
				const caller = createCaller(createAuthContext(ctx, sessionId));
				await expectTRPCError(
					() =>
						caller.procedure({
							receiptId: faker.string.uuid(),
							userId: "not-a-uuid",
						}),
					"BAD_REQUEST",
					`Zod error\n\nAt "userId": Invalid uuid`,
				);
			});
		});

		test("receipt does not exist", async ({ ctx }) => {
			const { sessionId, accountId } = await insertAccountWithSession(ctx);
			const caller = createCaller(createAuthContext(ctx, sessionId));
			await insertReceipt(ctx, accountId);
			const fakeReceiptId = faker.string.uuid();
			await expectTRPCError(
				() =>
					caller.procedure({
						receiptId: fakeReceiptId,
						userId: faker.string.uuid(),
					}),
				"NOT_FOUND",
				`Receipt "${fakeReceiptId}" does not exist.`,
			);
		});

		test("receipt is not owned by an account", async ({ ctx }) => {
			const { sessionId, accountId } = await insertAccountWithSession(ctx);
			await insertReceipt(ctx, accountId);

			const { id: foreignAccountId } = await insertAccount(ctx);
			const { id: foreignReceiptId } = await insertReceipt(
				ctx,
				foreignAccountId,
			);

			const caller = createCaller(createAuthContext(ctx, sessionId));
			await expectTRPCError(
				() =>
					caller.procedure({
						receiptId: foreignReceiptId,
						userId: faker.string.uuid(),
					}),
				"FORBIDDEN",
				`Not enough rights to remove participant from receipt "${foreignReceiptId}".`,
			);
		});

		describe("user", () => {
			test("does not exist", async ({ ctx }) => {
				const { sessionId, accountId } = await insertAccountWithSession(ctx);
				const { id: receiptId } = await insertReceipt(ctx, accountId);
				const fakeUserId = faker.string.uuid();

				const caller = createCaller(createAuthContext(ctx, sessionId));
				await expectTRPCError(
					() => caller.procedure({ receiptId, userId: fakeUserId }),
					"NOT_FOUND",
					`User "${fakeUserId}" does not exist.`,
				);
			});

			test("is not owned by the account", async ({ ctx }) => {
				const {
					sessionId,
					accountId,
					account: { email },
				} = await insertAccountWithSession(ctx);
				const { id: receiptId } = await insertReceipt(ctx, accountId);
				const { id: foreignAccountId } = await insertAccount(ctx);
				const { id: foreignUserId } = await insertUser(ctx, foreignAccountId);

				const caller = createCaller(createAuthContext(ctx, sessionId));
				await expectTRPCError(
					() => caller.procedure({ receiptId, userId: foreignUserId }),
					"FORBIDDEN",
					`User "${foreignUserId}" is not owned by "${email}".`,
				);
			});

			test("is not participating in the receipt", async ({ ctx }) => {
				const { sessionId, accountId } = await insertAccountWithSession(ctx);
				await insertReceipt(ctx, accountId);

				const { id: notParticipantUserId } = await insertUser(ctx, accountId);
				const { id: receiptId } = await insertReceipt(ctx, accountId);

				const caller = createCaller(createAuthContext(ctx, sessionId));
				await expectTRPCError(
					() =>
						caller.procedure({
							receiptId,
							userId: notParticipantUserId,
						}),
					"CONFLICT",
					`User "${notParticipantUserId}" does not participate in receipt "${receiptId}".`,
				);
			});
		});
	});

	describe("functionality", () => {
		test("user is removed", async ({ ctx }) => {
			const { sessionId, accountId } = await insertAccountWithSession(ctx);
			const { id: receiptId } = await insertReceipt(ctx, accountId);
			const { id: userId } = await insertUser(ctx, accountId);
			const { id: anotherUserId } = await insertUser(ctx, accountId);
			await insertReceiptParticipant(ctx, receiptId, userId, {
				resolved: true,
				role: "editor",
			});
			await insertReceiptParticipant(ctx, receiptId, anotherUserId);
			const { id: receiptItemId } = await insertReceiptItem(ctx, receiptId);
			await insertItemParticipant(ctx, receiptItemId, userId);
			await insertItemParticipant(ctx, receiptItemId, anotherUserId);

			// Verify unrelated data doesn't affect the result
			await insertReceiptItem(ctx, receiptId);
			const { id: anotherReceiptId } = await insertReceipt(ctx, accountId);
			await insertReceiptItem(ctx, anotherReceiptId);

			const caller = createCaller(createAuthContext(ctx, sessionId));
			await expectDatabaseDiffSnapshot(ctx, () =>
				caller.procedure({ receiptId, userId }),
			);
		});
	});
});
