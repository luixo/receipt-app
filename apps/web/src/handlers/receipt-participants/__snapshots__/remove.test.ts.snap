// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`receiptParticipants.remove > functionality > user is removed 1`] = `
{
  "deleted": {
    "itemParticipants": {
      "itemId:47c3aa72-d37e-4039-91bb-78a75ea71923|userId:4a950a61-6655-4a98-8554-dc797c4131a2": undefined,
    },
    "receiptParticipants": {
      "receiptId:ba0db3c1-5954-4d81-923a-e5ab58e56bfc|userId:4a950a61-6655-4a98-8554-dc797c4131a2": undefined,
    },
  },
}
`;
